Annex B.1.4 improvements for web-reality

anba opened the pull request:
Fixed:
- Removed invalid respecification of the PatternCharacter production (B.1.4 shouldn't affect Unicode RegExp)

Changes:
- Added InvalidBracedQuantifier to reject `/{1}/` (previously allowed through ExtendedTerm -> Atom -> PatternCharacter)
- Added ExtendedPatternCharacter to allow forms like `/{*/`, `/}*/` `/]/`
- ExtendedPatternCharacter also handles `/\c%/` by removing the `\` restriction (`/\c%/` is equivalent to writing `/\\c%/`)
- Similar changes for `\c%` have been added to "ClassAtomNoDash :: [~U] SourceCharacter ..." and "ClassAtomNoDashInRange :: SourceCharacter ..."
- Added support for `/[\c_]/` and `/[\c1]/` to ClassEscape (the extended forms `\c_` and `\c<decimal digit>` are only valid in CharacterClass)
- Changed "ClassEscape :: [~U] DecimalEscape" to allow `/[\8]/` by adding the restriction "but only if the integer value of DecimalEscape is 0" [1]
- Character ranges which start or end with a non-single element CharSet are now handled in a more web-conform way [2, 3]
- Merged Term and ExtendedTerm to avoid adding redundant semantics for ExtendedTerm
- Re-ordered some production rules for clarity and to match 21.2.1

[1] This change does not match JavaScriptCore (`/[\8]/.test('\\') == true` in JSC), but also see AtomEscape for the other case where `\8` is handled differently (`/\8/.test('8') == false` in JSC).
[2] `' -a'.split('').map(s => /[\s-a]/.test(s)) == [true, true, true]` in all major browsers. The previous Annex B semantics made `/[\s-t]/` equivalent to `/[s-t]/`.
[3] This change does not match Chakra (`/[\s-a-c]/.test('b') == true` in Chakra).

domenic commented:
Are all of these changes "bad" in that they should in theory only be applied to legacy environments? Or should some of them be in the main spec?

(In general the idea that someone will implement a JS regexp engine without implementing Annex B seems ludicrous, but that's a general Annex B problem and a separate battle we do not need to fight in this thread.)

anba commented:
`\c_` and `\c<decimal digit>` could be added to the main spec, because it's a relatively small and self-contained feature which shouldn't affect any current or future RegExp proposals. But almost nobody uses control escape sequences, so the Annex B dumping ground is good enough for now. :wink: 

The changes for `/\c%/` should definitely be confined to Annex B, because of their irregularities (`/\c%/` is treated like `/\\c%/` instead of `/c%/`).

Annex B allows dangling `]` and `}`, that's probably a more useful feature for the main spec. The same goes for the character range changes, except we need to decide on the exact semantics (follow Perl like Chakra does; use the semantics proposed in this PR (compatible to V8/JSC/SM); or even something entirely different).

anba commented:
Updated the PR per @goyakin review comments:
- Moved CharacterRangeOrUnion to the bottom of the annex section
- And restored the backslash character restriction in class classes.

bterlson commented:
Thanks for the updates @anba! This was hard work, I imagine, but a great help to have this behavior documented. A couple surprises in there for me :)

bterlson commented:
Committed as fbdfda6 (darn, so close to matching /[a-f]{7}/). Squashed into one commit and updated the commit message accordingly.
